### PR TITLE
Hotfix : Correct handling of error cube generation

### PR DESCRIPTION
**Pull request by rcooke-ast**
I propose that this goes straight to release, given that it directly affects the errors of the datacubes generated. We can also push this to develop, if you like?

A bug in the generation of the variance cube meant that error cubes were being underestimated. The fix implemented here ensures that all modes produce the correct residuals (see the image below).

top left — spec2d file
top right — datacube, NGP
bottom left — datacube, subpixel (3,3,3)
bottom right — datacube, subpixel (3,4,5)

The statistics are given in the subplot titles. Overall, this is acceptable to me. I have added a unit test, and I will add a VET test to the PR #1770 (and the corresponding dev-suite test). I'm running the tests now to be sure nothing else is broken.

image

### CODE REVIEW

**Review by profxj — APPROVED** [review]
all passes my muster

**Comment by rcooke-ast** [comment]
**I also generated a test to avoid this bug from creeping back in, in the future. It will be included in an unrelated** [dev suite PR](https://github.com/pypeit/PypeIt-development-suite/pull/311).

image

**Also, dev suite tests pass...**

image

File: pypeit/slittrace.py [hunk 1]
```diff
@@ -479,7 +479,7 @@ def get_slitlengths(self, initial=False, median=False):
         slitlen = right - left
         return np.median(slitlen, axis=1) if median else slitlen
 
-    def get_radec_image(self, wcs, alignSplines, tilts, slice_offset=None, initial=True, flexure=None):
+    def get_radec_image(self, wcs, alignSplines, tilts, slit_compute=None, slice_offset=None, initial=True, flexure=None):
```
Comment (kbwestfall) [inline pypeit/slittrace.py:482]: Add a description of `slit_compute` in the docstring.

File: pypeit/slittrace.py [hunk 2]
```diff
@@ -515,6 +515,15 @@ def get_radec_image(self, wcs, alignSplines, tilts, slice_offset=None, initial=T
...
             slits. Shape is (nslits, 2).
         """
+        # Check the input
+        if slit_compute is None:
+            # Compute all of the slits
+            slit_compute = np.arange(self.nslits)
+        elif isinstance(slit_compute, int):
+            slit_compute = np.atleast_1d(slit_compute)
+        elif isinstance(slit_compute, list):
+            slit_compute = np.array(slit_compute)
```
Comment (kbwestfall) [inline pypeit/slittrace.py]: I'm not sure if you need to differentiate between ints and lists here.  I.e., I think you get the same result if you use `np.atleast_1d` for both.

File: pypeit/core/datacube.py [hunk 3]
```diff
@@ -1633,6 +1668,7 @@ def subpixellate(output_wcs, bins, sciImg, ivarImg, waveImg, slitid_img_gpm, wgh
     # Prepare the output arrays
     outshape = (bins[0].size-1, bins[1].size-1, bins[2].size-1)
     binrng = [[bins[0][0], bins[0][-1]], [bins[1][0], bins[1][-1]], [bins[2][0], bins[2][-1]]]
+    binrng_arr = np.array(binrng)
```
Comment (kbwestfall) [inline pypeit/core/datacube.py]: Are you sure you need both `binrng` and `binrng_arr`?  I think `histrogramdd` should be able to accept a `numpy.array` for its `range` argument.

File: pypeit/core/datacube.py [hunk 4]
```diff
@@ -1655,45 +1691,54 @@ def subpixellate(output_wcs, bins, sciImg, ivarImg, waveImg, slitid_img_gpm, wgh
...
+                ra_corr, dec_corr = _all_dar[fr].correction( this_wave_subpix)
+            # Calculate spatial and spectral positions of the subpixels
+            spat_xx = np.add.outer(wpix[1], spat_x.flatten()).flatten()
+            spec_yy = np.add.outer(wpix[0], spec_y.flatten()).flatten()
+            # Transform this to spatial location
+            spatpos_subpix = _astrom_trans[fr].transform(sl, spat_xx, spec_yy)
+            spatpos = _astrom_trans[fr].transform(sl, wpix[1], wpix[0])
+            ssrt = np.argsort(spatpos)
+            # Initialize the voxel coordinates for each spec2D pixel
+            vox_coord = -1*np.ones((numpix, num_all_subpixels, 3))
```
Comment (kbwestfall) [inline pypeit/core/datacube.py]: `np.full((numpix, num_all_subpix, 3), -1, dtype=float)`

File: pypeit/core/datacube.py [hunk 5]
```diff
@@ -1705,13 +1750,29 @@ def subpixellate(output_wcs, bins, sciImg, ivarImg, waveImg, slitid_img_gpm, wgh
...
+                # Calculate the number of repeated indices for each subpixel - this is the subpixel weights
+                subpix_wght = np.apply_along_axis(utils.occurrences, 1, vox_index).flatten()
+            # Reshape the voxel coordinates
+            vox_coord = vox_coord.reshape(numpix * num_all_subpixels, 3)
+            # Use the "fast histogram" algorithm, that assumes regular bin spacing
+            flxcube += histogramdd(vox_coord, bins=outshape, range=binrng, weights=np.repeat(this_sci[this_sl] * this_wght_subpix[this_sl], num_all_subpixels) * subpix_wght)
+            varcube += histogramdd(vox_coord, bins=outshape, range=binrng, weights=np.repeat(this_var[this_sl] * this_wght_subpix[this_sl]**2, num_all_subpixels) * subpix_wght**3)
+            normcube += histogramdd(vox_coord, bins=outshape, range=binrng, weights=np.repeat(this_wght_subpix[this_sl], num_all_subpixels) * subpix_wght)
+            if debug:
+                # NOTE :: The residual cube does not include the subpixel weights - so, the following line is probably incorrect
```
Comment (kbwestfall) [inline pypeit/core/datacube.py]: How incorrect is it?  Is it better to comment it out?

File: pypeit/tests/test_utils.py [hunk 6]
```diff
@@ -138,3 +138,9 @@ def test_recursive_update():
     d = utils.recursive_update(d, u)
     assert sorted(list(d['rdx'].keys())) == ['detnum', 'spectrograph'], 'Missing merged keys'
 
+
+def test_occurrences():
+    inparr = np.array([1, 1, 4, 10, 5, 3, 5, 5, 8, 9])
+    tstarr = np.array([2, 2, 1,  1, 3, 1, 3, 3, 1, 1])
+    outarr = utils.occurrences(inparr)
+    assert np.array_equal(outarr, tstarr), 'Occurrences has failed'
```
Comment (kbwestfall) [inline pypeit/tests/test_utils.py:146]: Nice!

**Review by kbwestfall — APPROVED** [review]
Some nit-picks from me (as is my way).  Please include a comment in the change log (you'll need to add the 1.15.1dev.rst doc), and confirm you've updated the docs before merging.

File: pypeit/core/datacube.py (hunk 4 above)
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: Good call - thanks!

File: pypeit/core/datacube.py (hunk 5 above)
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: I think the debug can be removed now. I've added a test to monitor the residuals cubes.

File: pypeit/slittrace.py (hunk 1 above)
Comment (rcooke-ast) [reply pypeit/slittrace.py:482]: Good catch - thanks!

File: pypeit/slittrace.py (hunk 2 above)
Comment (rcooke-ast) [reply pypeit/slittrace.py]: Much cleaner 👍

File: pypeit/core/datacube.py (hunk 3 above)
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: You're quite right - thanks 👍  Shame on me for not realising that a numpy array can be used as an iterable... 🤦‍♂️

**Review by rcooke-ast — COMMENTED** [review]
Thanks @kbwestfall - all comments addressed. I will need to build docs shortly.

**Comment by rcooke-ast** [comment]
Docs are updated (with no warnings), and a `1.15.1dev` release doc has been added. I will run dev suite tests again to ensure that all tests pass before merging to release.

**Comment by rcooke-ast** [comment]
Tests pass...
image

**Pull request merged**